File: alex_report_code_2023-03-24/julia19/regression.jl
using Random
using Flux: Chain, Dense, Optimiser, ExpDecay,  train!, params, predict, mse
using Optimisers: SGD

function linear_regression(x, y, lr=0.01, epochs=100, batch_size=1)
    # Convert the input data and output data to tensors
    x = reshape(x, :, 1)
    y = reshape(y, :, 1)

    # Define the model
    model = Chain(Dense(1, 1))

    # Define the loss function
    loss(y_pred, y_true) = mse(y_pred, y_true)

    # Define the optimizer
    optimizer = Optimiser(ExpDecay(lr), SGD())

    # Train the model
    losses = Float32[]
    for epoch in 1:epochs
        for i in 1:batch_size:size(x, 1)
            x_batch = x[i:i+batch_size-1,:]
            y_batch = y[i:i+batch_size-1,:]

            Flux.train!(loss, params(model), [(x_batch, y_batch)], optimizer)
        end
        push!(losses, loss(predict(model, x), y))
    end

    # Return the trained model and the losses
    return model, losses
end

# Generate some sample input and output data
Random.seed!(123)
x = collect(range(0, stop=1, length=20))
y = 2*x .+ 1 .+ randn(20)*0.1

# Perform linear regression with batch size 5
model, losses = linear_regression(x, y, lr=0.01, epochs=100, batch_size=5)

# Print the learned parameters
println("Weight: $(model.layers[1].W[1])")
println("Bias: $(model.layers[1].b[1])")

# Convert x to a tensor for plotting
x_tensor = reshape(x, :, 1)

# Plot the data points and the learned line
using Plots
scatter(x, y, label="Data")
plot!(x, predict(model, x_tensor), label="Learned Line", color="red")
xlabel!("x")
ylabel!("y")
legend!()

# Calculate R-squared
y_pred = predict(model, x_tensor)
r_squared = 1 - sum((y .- y_pred).^2) / sum((y .- mean(y)).^2)
println("R-squared: $r_squared")

# Plot the loss curve
plot(losses, xlabel="Epoch", ylabel="Loss")

File: alex_report_code_2023-03-24/julia19/regression.py
# Create a code to perform a linear regression analysis
import numpy as np
import matplotlib.pyplot as plt
from scipy import stats

# Create a function to perform a linear regression analysis
def linear_regression(x, y):
    # Perform linear regression analysis
    slope, intercept, r_value, p_value, std_err = stats.linregress(x, y)
    # Print the results
    print('slope: %f, intercept: %f, r_value: %f, p_value: %f, std_err: %f' % (slope, intercept, r_value, p_value, std_err))
    # Plot the results
    plt.plot(x, y, 'o', label='original data')
    plt.plot(x, intercept + slope * x, 'r', label='fitted line')
    # add a plot title
    plt.title("Linear Regression")
    plt.legend()
    plt.show()

x = np.array([1, 2, 3, 4, 5, 6, 7, 8, 9, 10])
y = np.array([2, 4, 6, 8, 10, 12, 14, 16, 18, 20])
linear_regression(x, y)

# Output
# slope: 2.000000, intercept: 0.000000, r_value: 1.000000, p_value: 0.000000, std_err: 0.000000

# Duplicate the above code using sklearn
# Path: regression_sklearn.py
# Create a code to perform a linear regression analysis
from sklearn import linear_model
import numpy as np
import matplotlib.pyplot as plt

# Create a function to perform a linear regression analysis
def linear_regression(x, y):
    # Perform linear regression analysis
    regr = linear_model.LinearRegression()
    regr.fit(x, y)
    # Print the results
    print('slope: %f, intercept: %f' % (regr.coef_, regr.intercept_))
    # Plot the results
    plt.plot(x, y, 'o', label='original data')
    plt.plot(x, regr.intercept_ + regr.coef_ * x, 'r', label='fitted line')
    plt.legend()
    plt.show()

x = np.array([1, 2, 3, 4, 5, 6, 7, 8, 9, 10])
y = np.array([2, 4, 6, 8, 10, 12, 14, 16, 18, 20])
# linear_regression(x, y)
print("linear_regression(x, y) = ", linear_regression(x, y))

# repeat the above code using torch
# Path: regression_torch.py
# Create a code to perform a linear regression analysis
import torch
import torch.nn as nn

# Create a function to perform a linear regression analysis
def linear_regression_torch(x, y):
    # Perform linear regression analysis
    model = nn.Linear(1, 1)
    criterion = nn.MSELoss()
    optimizer = torch.optim.SGD(model.parameters(), lr=0.01)
    for epoch in range(1000):
        inputs = torch.from_numpy(x).float()
        labels = torch.from_numpy(y).float()
        optimizer.zero_grad()
        outputs = model(inputs)
        loss = criterion(outputs, labels)
        loss.backward()
        optimizer.step()
    # Print the results
    print('slope: %f, intercept: %f' % (model.weight.item(), model.bias.item()))
    # Plot the results
    plt.plot(x, y, 'o', label='original data')
    plt.plot(x, model.bias.item() + model.weight.item() * x, 'r', label='fitted line')
    # title
    plt.title("Linear Regression with torch")
    plt.legend()
    plt.show()

x = np.array([1, 2, 3, 4, 5, 6, 7, 8, 9, 10])
y = np.array([2, 4, 6, 8, 10, 12, 14, 16, 18, 20])
linear_regression_torch(x, y)
print("linear_regression_torch(x, y) = ", linear_regression_torch(x, y))

